kirin/piv/model_maker.py

```diff
@@ -515,4 +522,192 @@ def _request_navitia_stop_point(self, uic8):
+        )
+
+        # return result only if there are changes
+        if has_changes:
+            # update existing TripUpdate to avoid duplicates, keep created_at and preserve performance
+            res = db_trip_update if db_trip_update else new_trip_update
+            res.message = new_trip_update.message
+            res.status = new_trip_update.status
+            res.effect = new_trip_update.effect
+            res.physical_mode_id = new_trip_update.physical_mode_id
+            res.headsign = new_trip_update.headsign
+            res.stop_time_updates = res_stus
+            return res
+        else:
+            return None
+
+
+def list_stu_is_not_equal(left, right):
+    if len(right) != len(left):
+        return True
+    for i in range(0, len(left)):
+        if left[i].is_not_equal(right[i]):
+            return True
+    return False
+
+
+def find_enumerate_stu_in_stus(ref_stu, stus, start=0):
+    """
+    Find a stop_time in the navitia vehicle journey
+    :param ref_stu: the referent stop_time
+    :param stus: list of StopTimeUpdate available in a TripUpdate
+    :param start: order (comprised) to start the search from
+    :return: (order, stu) if found else (len(stus), None)
+    """
+    if start >= len(stus):
+        return len(stus), None
+
+    def same_stop(stu, ref_stu):
+        return stu.stop_id == ref_stu.stop_id
+
+    return next(((order, stu) for (order, stu) in enumerate(stus) if same_stop(stu, ref_stu)), (len(stus), None))
+
+
+def adjust_trip_update_consistency(trip_update, stus):
+    """
+    Adjust consistency of TripUpdate and StopTimeUpdates (side-effect).
+    NB : using list of StopTimeUpdate provided in stus, as the list is often not
+    attached to TripUpdate (to avoid ORM persiting it)
+    :param trip_update: TripUpdate to adjust (all but stop_time_updates)
+    :param stus: List of StopTimeUpdates to adjust
+    :return: None, just update given parameters
+    """
+    previous_stop_event_dt = datetime.datetime.min
+    for res_stu in stus:
+        # if trip is added: stops are either added or deleted (both can be for detour) + no delay
+        if trip_update.effect == TripEffect.ADDITIONAL_SERVICE.name:
+            if res_stu.arrival_status in SIMPLE_MODIF_STATUSES:
+                res_stu.arrival_status = ModificationType.add.name
+                res_stu.arrival_delay = datetime.timedelta(0)
+            if res_stu.departure_status in SIMPLE_MODIF_STATUSES:
+                res_stu.departure_status = ModificationType.add.name
+                res_stu.departure_delay = datetime.timedelta(0)
+
+        for stop_event_toggle in ["arrival", "departure"]:
+            # First fill missing datetime info
+            stop_event_dt = getattr(res_stu, stop_event_toggle)
+            if stop_event_dt is None:
+                # if info is missing, first consider the opposite event of same stop-time
+                opposite_event = {"arrival": "departure", "departure": "arrival"}
```

Comment:
How about creating a `opposite_event()` function to better express what you're doing with this `dict`?
```python
def opposite_event(event):
  if event == "departure"
    return "arrival"
  if event == "arrival"
    return "departure"
  raise ValueError("must be 'arrival' or 'departure': found '{}'".format(event))
```
or even something like the following would be more readable/logical in my opinion (you should be able to write this because we know `stop_event_toggle` is one of two possible values).
```suggestion
                opposite_event = "arrival" if stop_event_toggle == "departure" else "departure"
```

Actually, even an `enum` might be a more appropriate data structure for `arrival` and `departure`...

Reply:
Added and used an enum into https://github.com/CanalTP/kirin/pull/354/commits/3184888646146a94e877c842a40a69e1bded9179
Although the use should be widened as it's way better, but it's also widely used everywhere, so not in this fat PR.